### PR TITLE
Consul support as coordination service

### PR DESCRIPTION
**Pull request by juanjovazquez**
Proposal to incorporate Consul as other possible coordination service in addition to etcd.

### CODE REVIEW

File: constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpec.scala [hunk 1]
```diff
@@ -70,8 +70,8 @@ abstract class MultiNodeConstructrSpec extends MultiNodeSpec(ConstructrMultiNode
 
   "Constructr should manage an Akka cluster" in {
     runOn(nodes.head) {
-      "docker rm -f constructr-etcd".!(ProcessLogger(_ => ()))
-      s"""docker run --name constructr-etcd -d -p 2379:2379 quay.io/coreos/etcd:v2.2.1 -advertise-client-urls http://$host:2379 -listen-client-urls http://0.0.0.0:2379""".!
+      "docker-machine ssh default docker rm -f constructr-etcd".!(ProcessLogger(_ => ()))
```
Comment (hseeberger) [inline constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpec.scala]: What's the purpose of this change?

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ */
+
+package de.heikoseeberger.constructr.coordination
+
+import akka.http.scaladsl.client.RequestBuilding.{ Get, Put }
+import akka.http.scaladsl.model.StatusCodes.{ NotFound, OK }
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse, ResponseEntity, Uri }
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.Materializer
+
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Remove blank line.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  private val kvUri = Uri(s"http://$host:$port/v1/kv")
+
+  private val baseUri = kvUri.withPath(kvUri.path / "constructr" / prefix / clusterName)
+
+  private val nodesUri = baseUri.withPath(baseUri.path / "nodes")
+
+  private val createSessionUri = v1Uri.withPath(v1Uri.path / "session" / "create")
+
+  //private def renewSessionUri(sessionId: String) = kvUri.withPath(kvUri.path / "session" / "renew" / sessionId)
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: What does that comment mean? I recently collapsed AddingSelf and Refreshing, thereby entering AddingSelf repeatedly. So the entry in KV must still end after a TTL and hence be refreshed. Is this still the case here?

File: constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpec.scala (hunk 1 above)
Comment (juanjovazquez) [inline constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpec.scala]: We'll remove this change as we've seen that is not necessary.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala (hunk 3 above)
Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: This is precisely what we didn't understand. We're going to review again trying to understand it but I'm afraid that, when it comes to Consul, the procedure will have to be different as refreshing is not the same action as adding. In refreshing we'll have to refresh a `session` that we previously used to lock the key using a certain `sessionId`. That `sessionId` will have to be available in the `State Data`. We'll give thought to this part again and propose you a new version.

Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Hmm ... actually adding and refreshing isn't that different from my perspective: in both cases you have to write the same entry with a TTL. The only difference is that initially the entry didn't exist. Isn't it true that Consul needs a session adding and refreshing?

Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: The main difference with etcd in this aspect is that the ttl is not associated to the key but to the session. So, you need to associate a session to the key (via `acquire`) but the session is the one who owns the ttl. In order to extend the life of the key, we need to refresh this session, not the key, otherwise, the key would disappear when the session finishes.

Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: I know by now ;-)

So, what changes are needed to `Coordination`? Should we revert the collapsing change? How should the signatures look like? Should `addSelf` and `refresh` return some object (in case of Consul probably a session id)?

Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: To begin with, we'd need `addSelf` to return somehow the `sessionId` of the session that is linked to the new key. Following your design for `LockResult` should be something like this:
```scala
sealed trait AddingSelfResult
object AddingSelfResult {
  case class Success(sessionId: String) extends AddingSelfResult
  case object Failure extends AddingSelfResult
}
```
However, this design is not general enough as the `session` concept applies to Consul but doesn't apply to etcd. So, it could be something like this:
```scala
sealed trait AddingSelfResult[T]
object AddingSelfResult {
  case class Success[T](result: T) extends AddingSelfResult[T]
  case object Failure extends AddingSelfResult[Nothing]
}
```
Apart from that, we should be able to keep track the `sessionId` in the FSM State, so we could extend the `Data[A]` type constructor to incorporate that info. Could be sth as follows:
```scala
case class Data[A, B](nodes: List[A], coordinationRetriesLeft: Int, AddingSelfResult: B)
```
And, yes we should probably come back to the previous design where AddingSelf and Refreshing were separated.

WDYT?

Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: That looks pretty good. I'll revert the changes, because I think AddingSelf and Refreshing might need different failure handling. Thereby I'll apply the changes – slightly modified – you suggested. Stay tuned and watch far a respective PR from me.

Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Great!

File: constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala [hunk 4]
```diff
@@ -147,7 +147,11 @@ final class ConstructrMachine[A: Coordination.AddressSerialization, B <: Coordin
   }
 
   when(State.AddingSelf, coordinationTimeout) {
-    case Event(Coordination.SelfAdded(_), _) => goto(State.RefreshScheduled).using(stateData.copy(coordinationRetriesLeft = coordinationRetries))
+    case Event(Coordination.SelfAdded(context), _) =>
+      goto(State.RefreshScheduled).using(stateData.copy(
+        coordinationRetriesLeft = coordinationRetries,
+        context = context.asInstanceOf[B#Context]
```
Comment (juanjovazquez) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: This is ugly but we haven't found a cleaner version.

File: constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala [hunk 5]
```diff
@@ -158,6 +162,10 @@ final class ConstructrMachine[A: Coordination.AddressSerialization, B <: Coordin
 
   // Refreshing
 
+  onTransition {
+    case _ -> State.Refreshing => coordination.refresh(selfAddress, addOrRefreshTtl, stateData.context).pipeTo(self)
+  }
+
   when(State.Refreshing, coordinationTimeout) {
     case Event(Coordination.Refreshed(_), _) => goto(State.RefreshScheduled).using(stateData.copy(coordinationRetriesLeft = coordinationRetries))
   }
```
Comment (juanjovazquez) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: We think that this part is necessary in order to deal with the refreshing stage taking into account the backend context.

File: constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala (hunk 4 above)
Comment (hseeberger) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: ;-)
I think you should simply use a typed pattern in the pattern match three lines above: 

```
case Event(Coordination.SelfAdded(context: B#Context), _) => ...
```

File: constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala (hunk 5 above)
Comment (hseeberger) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: Oops, was that really missing? Thanks for catching! We need ways better tests to avoid such mistakes.

File: constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpecConsul.scala [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+class MultiNodeConstructrSpecConsulMultiJvmNode4 extends MultiNodeConstructrSpecConsul
+class MultiNodeConstructrSpecConsulMultiJvmNode5 extends MultiNodeConstructrSpecConsul
+
+abstract class MultiNodeConstructrSpecConsul[A: AddressSerialization] extends MultiNodeSpec(ConstructrMultiNodeConfigConsul)
+    with FreeSpecLike with Matchers with BeforeAndAfterAll {
+  import ConstructrMultiNodeConfig._
+  import RequestBuilding._
+  import system.dispatcher
+
+  def toNodes(s: String) = {
```
Comment (hseeberger) [inline constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpecConsul.scala]: Please move to line 138, i.e. after the tests (only vals should come before the tests).

File: constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpecConsul.scala [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+      implicitly[AddressSerialization[A]].fromBytes(decode(json.Key.as[String].substring("constructr/akka/MultiNodeConstructrSpecConsul/nodes/".length)))
+    }
+    Json.parse(s).as[List[Json]].map(jsonToNode)
+  }
+  implicit val mat = ActorMaterializer()
+
+  "Constructr should manage an Akka cluster" in {
+    runOn(nodes.head) {
+      "docker rm -f constructr-consul".!(ProcessLogger(_ => ()))
+      s"""docker run -d -p 8400:8400 -p 8500:8500 -p 8600:53/udp --name constructr-consul -h contructr-consul progrium/consul -server -bootstrap""".!
```
Comment (hseeberger) [inline constructr-akka/src/multi-jvm/scala/de/heikoseeberger/constructr/akka/MultiNodeConstructrSpecConsul.scala]: Why do you need the `-h ...`?

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  }
+
+  override def lock(ttl: Duration)(implicit ec: ExecutionContext, mat: Materializer): Future[LockResult] = {
+    // TODO Make idempotent, e.g. by first GETting and comparing the value, just PUTting, if not there with value = self
+    val uri = baseUri
+      .withPath(baseUri.path / "lock")
+
+    createSession(ttl)
+      .flatMap(sessionId => send(Put(uri.withQuery(Uri.Query("acquire" -> sessionId)))))
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Move `.withQuery(Uri.Query("acquire" -> sessionId))` up to line 64.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            .map(_.toBoolean)
+            .map(locked => if (locked) LockResult.Success else LockResult.Failure)
+        }
+        case HttpResponse(other, _, entity, _) =>
+          ignore(entity).map(_ => throw UnexpectedStatusCode(other))
+      }
+  }
+
+  override def addSelf[A: AddressSerialization](self: A, ttl: Duration)(implicit ec: ExecutionContext, mat: Materializer) =
+    (for {
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: This is hard to read. Use an intermediate result assigned to a val, e.g. `val responseAndSessionId = ...`, for the for-comprehension. Then call `flatMap` on that identifier.

**Comment by hseeberger** [comment]
Great stuff! I have a couple of minor style related changes. I assume the test passes? ;-)

**Comment by hseeberger** [comment]
May I also ask you to squash the various commits into one with a short and meaningful name, e.g. "Add support for Consul"?

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  private val nodesUri = baseUri.withPath(baseUri.path / "nodes")
+
+  private val createSessionUri = v1Uri.withPath(v1Uri.path / "session" / "create")
+
+  override def getNodes[A: AddressSerialization]()(implicit ec: ExecutionContext, mat: Materializer): Future[List[A]] = {
+    def unmarshalNodes(entity: ResponseEntity) = {
+      def toNodes(s: String) = {
+        import rapture.json._
+        import rapture.json.jsonBackends.spray._
+        def jsonToNode(json: Json) = {
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: More nitpicks ;-)
These curly braces aren't needed. Move the next line up, just after the equals sign.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+          implicitly[AddressSerialization[A]].fromBytes(decode(json.Key.as[String].substring(s"constructr/$prefix/$clusterName/nodes/".length)))
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Use `kvUri.path.toString` to avoid redundancy. See `EtcdCoordination`.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+      case HttpResponse(other, _, entity, _)    => ignore(entity).map(_ => throw UnexpectedStatusCode(other))
+    }
+  }
+
+  override def lock(ttl: Duration)(implicit ec: ExecutionContext, mat: Materializer): Future[LockResult] = {
+    val uri = baseUri
+      .withPath(baseUri.path / "lock")
+    val responseAndSession = for {
+      sessionId <- createSession(ttl)
+      response <- send(Put(uri.withQuery(Uri.Query("acquire" -> sessionId))))
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Move `.withQuery(Uri.Query("acquire" -> sessionId))` up to the definition of `uri`.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    } yield (response, sessionId)
+    responseAndSession.flatMap {
+      case (HttpResponse(OK, _, entity, _), _) => {
+        Unmarshal(entity).to[String]
+          .map(_.toBoolean)
```
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Can you try to directly unmarshal `to[Boolean]`?

**Comment by hseeberger** [comment]
Looks really good. Just some more nitpicks (sorry, I'm German, it's in my blood).

**Comment by juanjovazquez** [comment]
LOL No problem!. I'll fix this minor issues right away. Thanks!.

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala (hunk 13 above)
Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: We can do that but at a price. We need a compatible `Unmarshaller`in scope. I'd define this at the beginning of the class:
```scala
import akka.http.scaladsl.unmarshalling.PredefinedFromEntityUnmarshallers._
...
implicit def booleanUnmarshaller: FromEntityUnmarshaller[Boolean] = stringUnmarshaller.map(_.toBoolean)
```
What do you prefer?

Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: No, I was just hoping that there were a default `FromEntityUnmarshaller[Boolean]` in scope. Forget it ;-)

Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Wait, I didn't read the first line. So there is a predfined unmarshaller for `Boolean`? Then use it (via import).

Comment (juanjovazquez) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Unfortunately,  there isn't. The import brings an `stringUnmarshaller`. We'd have to add a `booleanUnmarshaller` mapping over the previous one. IMHO, it's not worth it.

File: constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala [hunk 14]
```diff
@@ -147,7 +147,11 @@ final class ConstructrMachine[A: Coordination.AddressSerialization, B <: Coordin
...
 
   when(State.AddingSelf, coordinationTimeout) {
-    case Event(Coordination.SelfAdded(_), _) => goto(State.RefreshScheduled).using(stateData.copy(coordinationRetriesLeft = coordinationRetries))
+    case Event(Coordination.SelfAdded(context: B#Context), _) =>
```
Comment (juanjovazquez) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: @hseeberger, this is an unchecked type pattern. We're facing a compile-time warning because of that:
```
Warning:(150, 50) abstract type pattern B#Context is unchecked since it is eliminated by erasure
    case Event(Coordination.SelfAdded(context: B#Context), _) =>
```
How do you prefer to deal with it?

File: constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala (hunk 13 above)
Comment (hseeberger) [inline constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala]: Ah, I misunderstood. I agree, keep the existing code.

File: constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala (hunk 14 above)
Comment (hseeberger) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: Add an `@unchecked` annotation, like above in line 113 `case Event(nodes: List[A] @unchecked, _) => ...`.

Comment (juanjovazquez) [inline constructr-machine/src/main/scala/de/heikoseeberger/constructr/machine/ConstructrMachine.scala]: Perfect.

**Comment by juanjovazquez** [comment]
@hseeberger, new installment trying to be more descriptive. Eliminated some duplications also.

**Comment by hseeberger** [comment]
This is awesome work. Thanks!
